src/server.c

```diff
@@ -516,6 +526,52 @@ uint64_t dictEncObjHash(const void *key) {
     }
 }
 
+static size_t kvstoreMetadataBytes(kvstore *kvs) {
+    UNUSED(kvs);
+    return sizeof(kvstoreMetadata);
+}
+
+static size_t kvstoreDictMetadataBytes(dict *d) {
+    UNUSED(d);
+    return sizeof(kvstoreDictMetadata);
+}
+
+static int kvstoreCanFreeDictIfNeeded(kvstore *kvs, int didx) {
+    kvstoreDictMetadata *meta = (kvstoreDictMetadata *)kvstoreGetDictMetadata(kvs, didx);
+#ifdef REDIS_TEST
+    serverAssert(meta->alloc_size == 0);
+#endif
+    /* Free if not in cluster */
+    if (!server.cluster_enabled) return 1;
+
+    if (server.cluster_slot_stats_enabled &&
+        (meta->cpu_usec || meta->network_bytes_in || meta->network_bytes_out) &&
+        clusterIsMySlot(didx))
+    {
+        /* Don't free if we have stats for this slot */
+        return 0;
+    }
+
+    /* Otherwise, we can free */
+    return 1;
+}
+
+static void kvstoreOnEmpty(kvstore *kvs, int is_release) {
+    kvstoreMetadata *meta = (kvstoreMetadata *)kvstoreGetMetadata(kvs);
+    if (!is_release)
+        memset(&meta->keysizes_hist, 0, sizeof(meta->keysizes_hist));
+}
+
+static void kvstoreOnDictEmpty(kvstore *kvs, int didx, int is_release) {
+    kvstoreDictMetadata *meta = (kvstoreDictMetadata *)kvstoreGetDictMetadata(kvs, didx);
+#ifdef REDIS_TEST
+    if (is_release) dictEmpty(kvstoreGetDict(kvs, didx), NULL);
+    serverAssert(meta->alloc_size == 0);
+#endif
+    if (!is_release)
+        memset(&meta->keysizes_hist, 0, sizeof(meta->keysizes_hist));
```

Comment:
I think we don't need to pass `is_release`. Just memset unconditionally. it is not that pain and not that often.